jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -14,4 +20,69 @@ public class JdbcTemplate {
     public JdbcTemplate(final DataSource dataSource) {
         this.dataSource = dataSource;
     }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper) {
+        return executeStatement(sql, (pstmt) -> executeQueryAndGet(pstmt, rowMapper));
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper, Object... values) {
+        return executeStatement(sql, (pstmt) -> {
+            bindValues(pstmt, values);
+            return executeQueryAndGet(pstmt, rowMapper);
+        });
+    }
+
+    private <T> T executeStatement(String sql, PreparedStatementCallBack<T> callBack) {
+        try (Connection conn = dataSource.getConnection(); PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            return callBack.callback(pstmt);
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private <T> List<T> executeQueryAndGet(PreparedStatement pstmt, RowMapper<T> rowMapper) throws SQLException {
+        List<T> result = new ArrayList<>();
+        try (ResultSet rs = pstmt.executeQuery()) {
+            while (rs.next()) {
+                result.add(rowMapper.mapRow(rs));
+            }
+        }
+        return result;
+    }
+
+    private void bindValues(PreparedStatement pstmt, Object... values) throws SQLException {
+        for (int idx = 1; idx <= values.length; idx++) {
+            pstmt.setObject(idx, values[idx - 1]);
+        }
```

Comment:
제안과 같이 작성하지 않은 이유가 궁금해요~!

```suggestion
        for (int idx = 0; idx <= values.length; idx++) {
            pstmt.setObject(idx, values[idx]);
        }
```

Reply:
```java
// PreparedStatement
void setObject(int parameterIndex, Object x) throws SQLException;
```

`parameterIndex`가 `1`부터 시작한다는 걸 명시적으로 표현하는 것이 명확하다 생각했어요. 아무래도 개발자라면 인덱스가 `0`부터 시작하는 것에 익숙하니깐요! `idx = 0`으로 시작하고 인자로 `idx+1`을 전달하는 것보다 `idx = 1`로 하고 `idx`를 넘기는 것이 명확하다고 생각해서요! 큰 차이는 없어요 :)